File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/LocationOverlayHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.LocationOverlay
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler.Companion.getterName
import io.flutter.plugin.common.MethodChannel

internal interface LocationOverlayHandler : OverlayHandler {
    fun handleLocationOverlay(
        overlay: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (overlay as LocationOverlay).let { l ->
        when (method) {
            anchorName -> setAnchor(l, arg!!)
            bearingName -> setBearing(l, arg!!)
            circleColorName -> setCircleColor(l, arg!!)
            circleOutlineColorName -> setCircleOutlineColor(l, arg!!)
            circleOutlineWidthName -> setCircleOutlineWidth(l, arg!!)
            circleRadiusName -> setCircleRadius(l, arg!!)
            iconName -> setIcon(l, arg!!)
            iconSizeName -> setIconSize(l, arg!!)
            iconAlphaName -> setIconAlpha(l, arg!!)
            positionName -> setPosition(l, arg!!)
            subAnchorName -> setSubAnchor(l, arg!!)
            subIconName -> setSubIcon(l, arg)
            subIconSizeName -> setSubIconSize(l, arg!!)
            subIconAlphaName -> setSubIconAlpha(l, arg!!)
            getterName(bearingName) -> getBearing(l, result::success)
            getterName(positionName) -> getPosition(l, result::success)
            else -> result.notImplemented()
        }
    }

    fun setAnchor(overlay: LocationOverlay, rawNPoint: Any)

    fun setBearing(overlay: LocationOverlay, rawBearing: Any)

    fun setCircleColor(overlay: LocationOverlay, rawColor: Any)

    fun setCircleOutlineColor(overlay: LocationOverlay, rawColor: Any)

    fun setCircleOutlineWidth(overlay: LocationOverlay, rawWidth: Any)

    fun setCircleRadius(overlay: LocationOverlay, rawRadius: Any)

    fun setIcon(overlay: LocationOverlay, rawNOverlayImage: Any)

    fun setIconSize(overlay: LocationOverlay, rawSize: Any)

    fun setIconAlpha(overlay: LocationOverlay, rawAlpha: Any)

    fun setPosition(overlay: LocationOverlay, rawLatLng: Any)

    fun setSubAnchor(overlay: LocationOverlay, rawNPoint: Any)

    fun setSubIcon(overlay: LocationOverlay, rawNOverlayImage: Any?)

    fun setSubIconSize(overlay: LocationOverlay, rawSize: Any)

    fun setSubIconAlpha(overlay: LocationOverlay, rawAlpha: Any)

    fun getBearing(overlay: LocationOverlay, success: (bearing: Float) -> Unit)

    fun getPosition(overlay: LocationOverlay, success: (latLng: Map<String, Any>) -> Unit)


    companion object {
        /* --- Messaging Name Define --- */
        private const val anchorName = "anchor"
        private const val bearingName = "bearing"
        private const val circleColorName = "circleColor"
        private const val circleOutlineColorName = "circleOutlineColor"
        private const val circleOutlineWidthName = "circleOutlineWidth"
        private const val circleRadiusName = "circleRadius"
        private const val iconName = "icon"
        private const val iconSizeName = "iconSize"
        private const val iconAlphaName = "iconAlpha"
        private const val positionName = "position"
        private const val subAnchorName = "subAnchor"
        private const val subIconName = "subIcon"
        private const val subIconSizeName = "subIconSize"
        private const val subIconAlphaName = "subIconAlpha"
    }
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/LocationOverlayHandler.swift
import NMapsMap
import Flutter

internal protocol LocationOverlayHandler: OverlayHandler {
    func setAnchor(_ overlay: NMFLocationOverlay, rawNPoint: Any)

    func setBearing(_ overlay: NMFLocationOverlay, rawBearing: Any)

    func setCircleColor(_ overlay: NMFLocationOverlay, rawColor: Any)

    func setCircleOutlineColor(_ overlay: NMFLocationOverlay, rawColor: Any)

    func setCircleOutlineWidth(_ overlay: NMFLocationOverlay, rawWidth: Any)

    func setCircleRadius(_ overlay: NMFLocationOverlay, rawRadius: Any)

    func setIcon(_ overlay: NMFLocationOverlay, rawNOverlayImage: Any)

    func setIconSize(_ overlay: NMFLocationOverlay, rawSize: Any)
    
    func setIconAlpha(_ overlay: NMFLocationOverlay, rawAlpha: Any)

    func setPosition(_ overlay: NMFLocationOverlay, rawLatLng: Any)

    func setSubAnchor(_ overlay: NMFLocationOverlay, rawNPoint: Any)

    func setSubIcon(_ overlay: NMFLocationOverlay, rawNOverlayImage: Any?)

    func setSubIconSize(_ overlay: NMFLocationOverlay, rawSize: Any)
    
    func setSubIconAlpha(_ overlay: NMFLocationOverlay, rawAlpha: Any)

    func getBearing(_ overlay: NMFLocationOverlay, success: (_ bearing: Double) -> Void)

    func getPosition(_ overlay: NMFLocationOverlay, success: (_ latLng: Dictionary<String, Any>) -> Void)
}

private let anchorName = "anchor"
private let bearingName = "bearing"
private let circleColorName = "circleColor"
private let circleOutlineColorName = "circleOutlineColor"
private let circleOutlineWidthName = "circleOutlineWidth"
private let circleRadiusName = "circleRadius"
private let iconName = "icon"
private let iconSizeName = "iconSize"
private let iconAlphaName = "iconAlpha"
private let positionName = "position"
private let subAnchorName = "subAnchor"
private let subIconName = "subIcon"
private let subIconSizeName = "subIconSize"
private let subIconAlphaName = "subIconAlpha"

internal extension  LocationOverlayHandler {
    func handleLocationOverlay(locationOverlay: NMFOverlay,
                               method: String,
                               args: Any?,
                               result: @escaping FlutterResult) {
        let locationOverlay = locationOverlay as! NMFLocationOverlay
        switch method {
        case anchorName: setAnchor(locationOverlay, rawNPoint: args!)
        case bearingName: setBearing(locationOverlay, rawBearing: args!)
        case circleColorName: setCircleColor(locationOverlay, rawColor: args!)
        case circleOutlineColorName: setCircleOutlineColor(locationOverlay, rawColor: args!)
        case circleOutlineWidthName: setCircleOutlineWidth(locationOverlay, rawWidth: args!)
        case circleRadiusName: setCircleRadius(locationOverlay, rawRadius: args!)
        case iconName: setIcon(locationOverlay, rawNOverlayImage: args!)
        case iconSizeName: setIconSize(locationOverlay, rawSize: args!)
        case iconAlphaName: setIconAlpha(locationOverlay, rawAlpha: args!)
        case positionName: setPosition(locationOverlay, rawLatLng: args!)
        case subAnchorName: setSubAnchor(locationOverlay, rawNPoint: args!)
        case subIconName: setSubIcon(locationOverlay, rawNOverlayImage: args)
        case subIconSizeName: setSubIconSize(locationOverlay, rawSize: args!)
        case subIconAlphaName: setSubIconAlpha(locationOverlay, rawAlpha: args!)
        case getterName(bearingName): getBearing(locationOverlay, success: result)
        case getterName(positionName): getPosition(locationOverlay, success: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
